lib/spack/docs/build_systems/luapackage.rst

```diff
@@ -0,0 +1,93 @@
+.. Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+   Spack Project Developers. See the top-level COPYRIGHT file for details.
+
+   SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+.. _luapackage:
+
+------------
+LuaPackage
+------------
+
+LuaPackage is a helper for the common case of Lua packages that provide
+a rockspec file.  This is not meant to take a rock archive, but to build
+a source archive or repository that provides a rockspec, which should cover
+most lua packages. In the case a Lua package builds by Make rather than
+luarocks, prefer MakefilePackage.
+
+^^^^^^
+Phases
+^^^^^^
+
+The ``LuaPackage`` base class comes with the following phases:
+
+#. ``preprocess`` - generate ninja files
+#. ``install`` - install the project
+
+By default, these phases run:
+
+.. code-block:: console
+
+   $ # preprocess is a noop by default
+   $ luarocks make .
+
+
+Any of these phases can be overridden in your package as necessary.
+
+^^^^^^^^^^^^^^^
+Important files
+^^^^^^^^^^^^^^^
+
+Packages that use the Lua/LuaRocks build system can be identified by the
+presence of a ``*.rockspec`` file. This file declares things
+like build instructions and dependencies.
+
+It is common for the rockspec file to list the lua version required in
+a dependency, the LuaPackage class adds appropriate dependencies on a Lua
+implementation, but it is a good idea to specify the version required with
+a ``depends_on`` statement.  The block normally will be a table definition like
+this:
+
+.. code-block:: lua
+
+   dependencies = {
+      "lua >= 5.1",
+   }
```

Comment:
What would be the equivalent syntax for Spack? Does we need to specify it for multiple possible providers at the moment?

Reply:
I suppose it would be something like:

```python
depends_on('lua@5.1:5.1.99', when='^lua')
```

It's only one because it's 5.1, and all the luajit variants are 5.1 variants.  If it were 5.3, only lua would be valid, so it would become:

```python
depends_on('lua@5.3')
```

That's actually a bit of an oddity with all this.  I considered putting a version on lua-lang, from a logical standpoint it should probably be there, but there's only one lua implementation that provides anything other than 5.1.

Reply:
I added some stuff to the docs, but in principle this should now be `depends_on('lua-lang@5.1:')`, as long as the package doesn't care what implementation it gets.